lib/schemes/refresh.js

```diff
@@ -0,0 +1,329 @@
+import defu from 'defu'
+import LocalScheme from './local'
+import { getProp } from '../utilities'
+
+export default class RefreshScheme extends LocalScheme {
+  constructor (auth, options) {
+    super(auth, defu(options, DEFAULTS))
+
+    this.refreshInterval = undefined
+    this.isRefreshing = false
+    this.failedRequestQueue = []
+  }
+
+  // ---------------------------------------------------------------
+  // Token Expiration helpers
+  // ---------------------------------------------------------------
+
+  _getExpiration () {
+    const _key = this.options.tokenExpirationPrefix + this.name
+
+    return this.$auth.$storage.getUniversal(_key)
+  }
+
+  _setExpiration (token) {
+    const _key = this.options.tokenExpirationPrefix + this.name
+
+    return this.$auth.$storage.setUniversal(_key, token)
+  }
+
+  _syncExpiration () {
+    const _key = this.options.tokenExpirationPrefix + this.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  // ---------------------------------------------------------------
+  // Refresh token helpers
+  // ---------------------------------------------------------------
+
+  _getRefreshToken () {
+    const _key = this.options.refreshTokenPrefix + this.name
+
+    return this.$auth.$storage.getUniversal(_key)
+  }
+
+  _setRefreshToken (refreshToken) {
+    const _key = this.options.refreshTokenPrefix + this.name
+
+    return this.$auth.$storage.setUniversal(_key, refreshToken)
+  }
+
+  _syncRefreshToken () {
+    const _key = this.options.refreshTokenPrefix + this.name
+
+    return this.$auth.$storage.syncUniversal(_key)
+  }
+
+  async _updateTokens (result) {
+    if (this.options.tokenRequired) {
+      let token = getProp(result, this.options.endpoints.login.propertyName)
+      if (this.options.tokenType) {
+        token = this.options.tokenType + ' ' + token
+      }
+
+      // Update access token
+      this.$auth.setToken(this.name, token)
+      super._setToken(token)
+
+      // Update refresh token and register refresh-logic with axios
+      const refreshToken = getProp(result, this.options.endpoints.refresh.propertyName)
+      if (refreshToken !== undefined) {
+        this._setRefreshToken(refreshToken)
+
+        const _tokenCreatedAt = getProp(result, this.options.createdAtProperty) || Date.now()
+        const _tokenTTL = getProp(result, this.options.expiresInProperty) || this.options.defaultExpirationTime
+        const tokenExpiration = _tokenCreatedAt + (_tokenTTL * 1000)
+        this._setExpiration(tokenExpiration)
+      }
+
+      // Update client id
+      const clientId = getProp(result, this.options.clientIdProperty)
+      if (clientId) {
+        this.$auth.setClientId(this.name, clientId)
+      }
+
+      // If have failed requests in the queue
+      if (this.failedRequestQueue.length) {
```

Comment:
This was added because the cause of the error was the "refresh". So if requests conflicted with refresh request, we resend them. Otherwise, everyone would need to make their own "resend" if token is invalid.